Overhaul OS documentation

Mickeon opened the pull request:
~Draft because more could be done soon. I would love your assistance!~

Continuation of [similar past efforts](https://github.com/godotengine/godot/pulls?q=is%3Apr+author%3AMickeon+overhaul) with the same goal of refreshing the documentation up a notch.

This is particularly prickly due of the sheer size. There's also a constant fear of possibly losing important information, or making use of incorrect technical terms. A lot of the time is spent reading the original documentation and looking at the source code, too. Fortunately, a decent majority of it is written... quite up to date, too. 

- Made the wording match how other classes are documented A LOT more closely;
   - Simplified wording which may be seen as awkward to read, _kinda like this very sentence_;
   - Use the present simple over future tense when applicable (`will return` -> `returns`);
   - Turned some long-winded sentences into notes at the end (and viceversa);
   - Turned some multiple paragraphs into lists;
- Link to other methods and parameters more often;
- Add more examples all around;
- Describe ERR_* constants return values;
- Note more methods moved out of the OS class in the transition to Godot 4;
- All **SystemDir** constants' descriptions have been prefixed with "Refers to":
    - They could be mistaken to be constant strings, otherwise.

------

### **`close_midi_inputs`, `get_connected_midi_inputs`, `open_midi_inputs`**

- Link the methods between one another.

### `get_cmdline_user_args`
- Rework description entirely.
   - The previous one was a bit unorganized.

### `get_keycode_string`, `find_keycode_from_string`
- Add considerably more examples.

### `request_permission`, `revoke_granted_permissions`
- Rework description entirely.
   - The previous one  was absolutely not formatted the same as the rest of the documentation.
       
### `get_user_data_dir`
- Mention `ProjectSettings.application/config/use_custom_user_dir` directly.

------------------------------------

Feedback is very, very welcome, even in the draft state. And especially right now. This class is huge and very technical.

In fact, I would like to propose a few questions that may help with this cause:
- [ ] "Web platform" or "Web platforms"? Capitalized "Web" or not?
- [ ] Is `OS.move_to_trash()` really implemented only on Android, Linux, macOS and Windows?
- [ ] Do `OS.request_permission` and `OS.request_permission` return `true` when permissions are granted successfully?
- [x] ~What does `OS.set_use_file_access_save_and_swap` really do?~

Mickeon committed:
Overhaul OS documentation

bruvzg commented on doc/classes/OS.xml:
```diff
@@ -39,6 +40,7 @@
 				Creates a new instance of Godot that runs independently. The [param arguments] are used in the given order and separated by a space.
 				If the process creation succeeds, the method will return the new process ID, which you can use to monitor the process (and potentially terminate it with [method kill]). If the process creation fails, the method will return [code]-1[/code].
+				See [method create_process] if you wish to run a different process.
 				[b]Note:[/b] This method is implemented on Android, iOS, Linux, macOS and Windows.
```
```suggestion
				[b]Note:[/b] This method is implemented on Android, Linux, macOS and Windows.
```
None of the process creation methods are available on non-jailbroken iOS.

Mickeon replied:
Oops. This removal was likely accidental.

Mickeon replied:
Huh. Heh. I didn't touch that one. Glad these PRs force people to pay attention

bruvzg commented on doc/classes/OS.xml:
```diff
@@ -91,10 +93,11 @@
 			<param index="4" name="open_console" type="bool" default="false" />
 			<description>
-				Executes a command. The file specified in [param path] must exist and be executable. Platform path resolution will be used. The [param arguments] are used in the given order, separated by spaces, and wrapped in quotes. If an [param output] [Array] is provided, the complete shell output of the process will be appended as a single [String] element in [param output]. If [param read_stderr] is [c ...
+				Executes the given command. The file specified in [param path] must exist and be executable. Platform path resolution will be used. The [param arguments] are used in the given order, separated by spaces, and wrapped in quotes.
```
> Executes the given command.

It won't execute built-in shell commands or scripts, so the description is misleading. The only difference from `create_process`, this method is blocking and reads output.

bruvzg commented on doc/classes/OS.xml:
```diff
@@ -176,91 +195,107 @@
 			<description>
-				Returns the value of an environment variable. Returns an empty string if the environment variable doesn't exist.
+				Returns the value of the given environment variable, or an empty string if [param variable] doesn't exist.
 				[b]Note:[/b] Double-check the casing of [param variable]. Environment variable names are case-sensitive on all platforms except Windows.
```
It might be worth adding a note that on macOS, apps do not have access to the shell environment variables.

bruvzg commented:
> What does OS.set_use_file_access_save_and_swap really do?

If enabled when opening a file for writing, a temporary file is used first, and only when it's closed it is renamed or used to replace an existing file. This is useful in case the original file is opened by other app, scanned by antivirus for example.

Mickeon commented:
Pushed a commit that also attempts to address the above comments. I hope you check again.

Mickeon commented:
How did the latest unit test even fail?

YuriSizov commented:
> How did the latest unit test even fail?

Unit tests didn't fail, but there is a problem with thread sanitizer runs. It's not related to your PR and seems to happen sporadically, a re-run fixed it.

Mickeon commented:
Rebased after https://github.com/godotengine/godot/pull/81416

Mickeon commented on doc/classes/OS.xml:
```diff
@@ -128,13 +131,29 @@
+				print(OS.find_keycode_from_string("Unknown"))   # Prints 0 (KEY_NONE)
+				[/gdscript]
+				[csharp]
+				GD.Print(OS.FindKeycodeFromString("C"));         // Prints 67 (KEY_C)
```
But how does that work with `GD.Print(OS.FindKeycodeFromString("Shift+Tab"))`? Is the printed value a raw number or does C# infer it's a combination?

raulsntos replied:
Right, since there's no value in the enum that corresponds to `37748738`, it will just print the number.

Mickeon replied:
Alright, so this is correct. Good stuff.

Mickeon commented:
Pushed changes that accept the C# suggestions above. Thank you @raulsntos for the additional attention to this PR.

raulsntos commented on doc/classes/OS.xml:
```diff
@@ -176,91 +195,108 @@
 			<description>
-				Returns an array of MIDI device names.
-				The returned array will be empty if the system MIDI driver has not previously been initialized with [method open_midi_inputs].
+				Returns an array of MIDI device names. The returned array will be empty if the system MIDI driver has not previously been initialized with [method open_midi_inputs]. See also [method close_midi_inputs].  
```
There are 2 spaces at the end of this line.

```suggestion
				Returns an array of MIDI device names. The returned array will be empty if the system MIDI driver has not previously been initialized with [method open_midi_inputs]. See also [method close_midi_inputs].
```

Mickeon replied:
Interesting. The checks didn't automatically detect this.

Deozaan commented on doc/classes/OS.xml:
```diff
@@ -48,9 +50,9 @@
-				If the process creation succeeds, the method will return the new process ID, which you can use to monitor the process (and potentially terminate it with [method kill]). If the process creation fails, the method will return [code]-1[/code].
+				Creates a new process that runs independently of Godot. It will not terminate when Godot terminates. The path specified in [param path] must exist and be executable file or macOS .app bundle. Platform path resolution will be used. The [param arguments] are used in the given order and separated by a space.
+				On Windows, if [param open_console] is [code]true[/code] and the process is a console app, a new terminal window will be opened.
+				If the process is successfully created, this method returns the its process ID, which you can use to monitor the process (and potentially terminate it with [method kill]). Otherwise this method returns [code]-1[/code].
```
"returns the its process ID"

Mickeon commented:
Rebased after:
- https://github.com/godotengine/godot/pull/82691
- https://github.com/godotengine/godot/pull/82441

AThousandShips commented on doc/classes/OS.xml:
```diff
@@ -16,30 +16,32 @@
 			<description>
-				Crashes the engine (or the editor if called within a [code]@tool[/code] script). This should [i]only[/i] be used for testing the system's crash handler, not for any other purpose. For general error reporting, use (in order of preference) [method @GDScript.assert], [method @GlobalScope.push_error] or [method alert]. See also [method kill].
+				Crashes the engine (or the editor if called within a [code]@tool[/code] script).
+				[b]Note:[/b] This method should [i]only[/i] be used for testing the system's crash handler, not for any other purpose. For general error reporting, use (in order of preference) [method @GDScript.assert], [method @GlobalScope.push_error] or [method alert]. See also [method kill].
```
```suggestion
				[b]Note:[/b] This method should [i]only[/i] be used for testing the system's crash handler, not for any other purpose. For general error reporting, use (in order of preference) [method @GDScript.assert], [method @GlobalScope.push_error], or [method alert]. See also [method kill].
```
Please use [Oxford comma](https://docs.godotengine.org/en/latest/contributing/documentation/docs_writing_guidelines.html#use-the-oxford-comma-to-enumerate-anything).

AThousandShips commented on doc/classes/OS.xml:
```diff
@@ -176,91 +195,108 @@
+				GD.Print(OS.GetKeycodeString((Key)KeyModifierMask.MaskShift | Key.Tab)); // Prints "Shift+Tab"
+				[/csharp]
+				[/codeblocks]
+				See also [method find_keycode_from_string], [member InputEventKey.keycode] and [method InputEventKey.get_keycode_with_modifiers].
```
```suggestion
				See also [method find_keycode_from_string], [member InputEventKey.keycode], and [method InputEventKey.get_keycode_with_modifiers].
```

AThousandShips commented on doc/classes/OS.xml:
```diff
@@ -4,8 +4,8 @@
-		This class wraps the most common functionalities for communicating with the host operating system, such as the video driver, delays, environment variables, execution of binaries, command line, etc.
-		[b]Note:[/b] In Godot 4, [OS] functions related to window management were moved to the [DisplayServer] singleton.
+		The [OS] class wraps the most common functionalities for communicating with the host operating system, such as the video driver, delays, environment variables, execution of binaries, command line, etc.
+		[b]Note:[/b] In Godot 4, [OS] functions related to window management, clipboard and TTS were moved to the [DisplayServer] singleton (and the [Window] class). Functions related to time were removed and are only available in the [Time] class.
```
```suggestion
		[b]Note:[/b] In Godot 4, [OS] functions related to window management, clipboard, and TTS were moved to the [DisplayServer] singleton (and the [Window] class). Functions related to time were removed and are only available in the [Time] class.
```

AThousandShips commented on doc/classes/OS.xml:
```diff
@@ -423,7 +461,7 @@
 			<param index="6" name="italic" type="bool" default="false" />
 			<description>
-				Returns an array of the system substitute font file paths, which are similar to the font with [param font_name] and style for the specified text, locale and script. Returns empty array if no matching fonts found.
+				Returns an array of the system substitute font file paths, which are similar to the font with [param font_name] and style for the specified text, locale and script. Returns an empty array if no matching fonts found.
```
```suggestion
				Returns an array of the system substitute font file paths, which are similar to the font with [param font_name] and style for the specified text, locale, and script. Returns an empty array if no matching fonts found.
```

AThousandShips reviewed (approved):
Some minor changes, style looks good otherwise, can't speak for the factual correctness however so someone else needs to verify

Mickeon commented:
Addressed the above feedback. Thank you.

I also moved the mention of `kill` in the `crash` method description outside of the note, because it is directly related.

akien-mga commented on doc/classes/OS.xml:
```diff
@@ -297,30 +334,30 @@
 				match OS.get_name():
-				    "Windows":
-				        print("Windows")
+				    "Windows", "UWP":
```
We removed the UWP platform.

akien-mga commented on doc/classes/OS.xml:
```diff
@@ -297,30 +334,30 @@
 				    "Android":
-				        print("Android")
+				        print("This is an Android!")
 				    "iOS":
-				        print("iOS")
+				        print("This is an iOS!")
```
Any specific reason why Android and iOS get a different pattern than the others?

Mickeon replied:
No, not particularly, in hindsight. Perhaps "Welcome to iOS" sounded off at first.

akien-mga commented on doc/classes/OS.xml:
```diff
@@ -500,30 +538,42 @@
 				Returns [code]true[/code] if the feature for the given feature tag is supported in the currently running instance, depending on the platform, build, etc. Can be used to check whether you're currently running a debug build, on a certain platform or arch, etc. Refer to the [url=$DOCS_URL/tutorials/export/feature_tags.html]Feature Tags[/url] documentation for more details.
 				[b]Note:[/b] Tag names are case-sensitive.
 				[b]Note:[/b] On the web platform, one of the following additional tags is defined to indicate host platform: [code]web_android[/code], [code]web_ios[/code], [code]web_linuxbsd[/code], [code]web_macos[/code], or [code]web_windows[/code].
```
```suggestion
				[b]Note:[/b] On the Web platform, one of the following additional tags is defined to indicate host platform: [code]web_android[/code], [code]web_ios[/code], [code]web_linuxbsd[/code], [code]web_macos[/code], or [code]web_windows[/code].
```

akien-mga commented on doc/classes/OS.xml:
```diff
@@ -578,12 +628,14 @@
 				OS.MoveToTrash(ProjectSettings.GlobalizePath(fileToRemove));
 				[/csharp]
 				[/codeblocks]
+				[b]Note:[/b] This method is implemented on Android, Linux, macOS and Windows.
```
Oxford comma.
```suggestion
				[b]Note:[/b] This method is implemented on Android, Linux, macOS, and Windows.
```

akien-mga commented on doc/classes/OS.xml:
```diff
@@ -638,25 +691,26 @@
 			<description>
-				Enables backup saves if [param enabled] is [code]true[/code].
+				If [param enabled] is [code]true[/code], when opening a file for writing, a temporary file is used in its place. When closed, it is automatically applied to the target file.
+				This can useful when files may be opened by other applications, such as antiviruses, text editors, or even the Godot Editor itself.
```
```suggestion
				This can useful when files may be opened by other applications, such as antiviruses, text editors, or even the Godot editor itself.
```

akien-mga commented on doc/classes/OS.xml:
```diff
@@ -666,30 +720,33 @@
-				[b]Note:[/b] Currently this method is only implemented on Windows and macOS. On other platforms, it will fallback to [method shell_open] with a directory path of [param file_or_dir_path] with prefix [code]file://[/code].
+				Requests the OS to open the file manager, navigate to the given [param file_or_dir_path] and select the target file or folder.
+				If [param open_folder] is [code]true[/code] and [param file_or_dir_path] is a valid directory path, the OS will open the file manager and navigate to the target folder without selecting anything.
+				Use [method ProjectSettings.globalize_path] to convert a [code]res://[/code] or [code]user://[/code] project path into a system path to use for with this method.
```
```suggestion
				Use [method ProjectSettings.globalize_path] to convert a [code]res://[/code] or [code]user://[/code] project path into a system path to use with this method.
```

akien-mga reviewed (approved):
Only skimmed through the changes, looks good overall.

akien-mga commented on doc/classes/OS.xml:
```diff
@@ -578,12 +628,14 @@
 				OS.MoveToTrash(ProjectSettings.GlobalizePath(fileToRemove));
 				[/csharp]
 				[/codeblocks]
+				[b]Note:[/b] This method is implemented on Android, Linux, macOS, and Windows.
```
Actually I realize that the other similar statements don't use the Oxford comma in this file. Maybe it's best to remove it for consistency.

Then a unification pass could be done in another PR as we currently use a mix of both:
```
$ rg -g'*.xml' "is implemented on .*, and [A-Za-z]+\." | wc -l
18
$ rg -g'*.xml' "is implemented on .*[^,] and [A-Za-z]+\." | wc -l
39
```

18 with Oxford comma, 39 without.

Mickeon replied:
Okiedokie then, unless we want to explicitly make an exception when listing platform availability.

Mickeon commented:
Reverted that comma, but also updated the `midi` methods' descriptions taking them from https://github.com/godotengine/godot/pull/86693. So, whichever gets merged first, they're the same.

akien-mga commented:
Thanks!
